### PR TITLE
Use DB replica(s) on GET requests

### PR DESCRIPTION
**Pull request by dpoirier**
but send POST, PUT requests to the master

I've tried registering, logging in and out,
submitting questions, and merging, and they
all seem to work with this enabled.

(This also adds a 'dan' environment that I used to test this.
That can be broken out into another pull, or removed entirely,
if the team prefers.)

### CODE REVIEW

**Comment by dpoirier** [comment]
Working on adding tests for that router.

**Comment by dpoirier** [comment]
Tests added.

**Comment by dpoirier** [comment]
Tobias asked earlier: Any reason not to use django-balancer mixins, to avoid having to do the round-robin bits yourself?

django-balancer's RoundRobinMasterSlaveRouter looks at each query to see if it's a read or write. You had mentioned possible problems if in a view, we made a db change and then didn't see the change when reading afterward, so I wanted to have each request use the same database throughout.

There's also PinningWMSRouter, which is closer, but doesn't start pinning until it sees the first write query, so a view that had some read queries before its first write query might see inconsistent data.

I did borrow django-balancer's mechanism for remembering whether to use the master db during a request by using thread-local variables, since the db router doesn't have access to the request.

The one thing the PinningWMSRouter does that this doesn't is continue using the master database for a while for all requests in the session, even subsequent GETs.  If that seems important, we could add it.

**Comment by tobiasmcnulty** [comment]
I'm intrigued. I like this approach. I think pinning the session may be required for any post/redirect/get pattern -- when under load the slave can get several seconds (if not more) behind, likely longer than it would take for the redirect & GET to complete.

Any reason not to use the router in settings.py?

**Comment by dpoirier** [comment]
Updated to pin sessions to the master DB for a while when they write to the DB. Also some refactoring, and tried to get rid of the term 'slave'.

File: opendebates/tasks.py [hunk 1]
```diff
@@ -19,19 +20,28 @@ def update_recent_events():
...
-    votes = Vote.objects.select_related(
-        "voter", "voter__user", "submission").exclude(
-        voter=F('submission__voter')).order_by("-id")[:10]
-    submissions = Submission.objects.select_related(
-        "voter", "voter__user").order_by("-id")[:10]
+    try:
+        # No middleware on tasks, so this won't get set otherwise.
+        # Tell the DB router this thread only needs to read the DB, not write.
+        # We run this task frequently, so it's worthwhile.
+        set_thread_readonly_db()
```
Comment (tobiasmcnulty) [inline opendebates/tasks.py]: If/when this needs to be done elsewhere, might be nice to refactor this to use context manager or decorator syntax.

**Comment by tobiasmcnulty** [comment]
Looks solid to me, and I'm okay with keeping the 'dan' environment around if others are.

:shipit:

**Pull request merged**